### PR TITLE
chore: enhance client-related typings

### PR DESCRIPTION
**Pull request by kyle221b**
# Fixes #

The MailService has a method to set a client, but it isn't exposed in typings. Adding this so there is an easier way to customize underlying agents used (for keep-alive and other connection-pooling settings) without having compilation problems (or subverting the type system by just using 'any')

### Checklist
- [x] I acknowledge that all my contributions will be made under the project's license
- [X] I have made a material change to the repo (functionality, testing, spelling, grammar)
- [X] I have read the [Contribution Guidelines](https://github.com/sendgrid/sendgrid-nodejs/blob/main/CONTRIBUTING.md) and my PR follows them
- [X] I have titled the PR appropriately
- [X] I have updated my branch with the main branch
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added the necessary documentation about the functionality in the appropriate .md file
- [ ] I have added inline documentation to the code I modified

If you have questions, please file a [support ticket](https://support.sendgrid.com), or create a GitHub Issue in this repository.

### CODE REVIEW

File: packages/client/src/client.d.ts [hunk 1]
```diff
@@ -46,5 +46,8 @@ declare class Client {
   request(data: ClientRequest, cb?: (err: ResponseError, response: [ClientResponse, any]) => void): Promise<[ClientResponse, any]>;
 }
 
-declare const client: Client & { Client: typeof Client };
+declare const client: Client;
+// @ts-ignore
```
Comment (kyle221b) [inline packages/client/src/client.d.ts:50]: This matches style of mail.d.ts file. Part that I wanted to change here mainly was the `{ Client: typeof Client }` piece. I get the feeling that the goal there was to expose the class as a type, but we do `typeof` on the class itself, instead of an instance of it, so it ends up being a type with just a `prototype` field. Let me know if I'm missing something here.

**Comment by kyle221b** [comment]
@thinkingserious any chance you or someone from your team could take a look at this?

File: packages/client/src/client.d.ts (hunk 1 above)
Comment (thinkingserious) [reply packages/client/src/client.d.ts:50]: I believe this change might break folks: https://github.com/sendgrid/sendgrid-nodejs/issues/986

File: packages/client/src/client.d.ts [hunk 2]
```diff
@@ -46,5 +46,8 @@ declare class Client {
...
 export = client
+
+export {Client};
```
Comment (thinkingserious) [inline packages/client/src/client.d.ts:53]: We would need to revert this change also if we move back to `declare const client: Client & { Client: typeof Client };`

File: packages/helpers/classes/request.d.ts [hunk 3]
```diff
@@ -7,4 +9,5 @@ export default interface RequestOptions<TData = any, TParams = object> {
     qs?: TParams;
     body?: TData;
     headers?: object;
+    httpsAgent?: https.Agent;
```
Comment (thinkingserious) [inline packages/helpers/classes/request.d.ts:12]: I believe this data is stored in `headers`.

**Review by thinkingserious — COMMENTED** [review]
Thank you for the PR @kyle221b!

File: packages/client/src/client.d.ts (hunk 1 above)
Comment (kyle221b) [reply packages/client/src/client.d.ts:50]: yeah the goal here was to export `Client` as not just a value with type `typeof Client`, but to export the class itself, which could be used as a value or a type. im pretty sure this will only enhance the existing functionality. currently, users could do:
`const a = new Client()`

but the only way they could pass the actual type around is by doing a default import of the singleton client and then doing a `typeof` on it.

am i missing something here?

File: packages/client/src/client.d.ts (hunk 2 above)
Comment (kyle221b) [reply packages/client/src/client.d.ts:53]: left a comment above about why I did this. let me know what you think

File: packages/helpers/classes/request.d.ts (hunk 3 above)
Comment (kyle221b) [reply packages/helpers/classes/request.d.ts:12]: this is a top-level field on an `axios` request. i don't think we'd want to pass it in headers b/c it won't actually get sent across the wire, it's just connection settings essentially.

if we were to add it to headers, then there would need to be a code change in the JS file to filter it out and pass it as a top-level field in the axios request

File: packages/client/src/client.d.ts (hunk 1 above)
Comment (thinkingserious) [reply packages/client/src/client.d.ts:50]: I don't quite understand why this method is better than what we have. Perhaps a good way to illustrate is to add a sample quick start [here](https://github.com/sendgrid/sendgrid-nodejs/tree/main/docs/use-cases).

Comment (kyle221b) [reply packages/client/src/client.d.ts:50]: To give a quick code example:

```
import { Client } from '@sendgrid/client';

// This is possible in both versions
const client = new Client();

// This (I think) is only possible in the new version since Client is only a value in the old version and not a type
function somethingThatTakesInAClient(client: Client) {
...
}
```

Comment (kyle221b) [reply packages/client/src/client.d.ts:50]: while it isn't necessary, you do it for most of your other types that I've seen, and if you are going to export functionality, it seems to me it would also make sense to export the type

Comment (kyle221b) [reply packages/client/src/client.d.ts:50]: to further illustrate, in the current version, we could accomplish the same by doing:
```
// Note we are importing the singleton client exported by the lib as well as the class itself (as a value)
import client, { Client } from '@sendgrid/client';

// We can do a typeof on the singleton instance. This is equivalent to what is in the last example. For this one you need // to import a singleton instance of the client just to get the type
function somethingThatTakesInAClient(arg: typeof client) {
...
}

// This compiles, but the behavior isn't what you expect. Since we are doing typeof on the class, the only members are // things on the function prototype, so we can't access methods like 'setApiKey' for example
function anotherThingThatLooksLikeItTakesInAClientButDoesntReally(arg: typeof Client) {
...
}
```

Comment (kyle221b) [reply packages/client/src/client.d.ts:50]: Here's an example where you use the proposed pattern: https://github.com/sendgrid/sendgrid-nodejs/blob/main/packages/mail/src/mail.d.ts#L39-L42.

**Comment by thinkingserious** [comment]
@kyle221b,

Your reasoning on [this thread](https://github.com/sendgrid/sendgrid-nodejs/pull/1227#pullrequestreview-545387327) makes sense to me. However, I'm still concerned about the points made in [this PR](https://github.com/sendgrid/sendgrid-nodejs/issues/986).

@danhlee329, @dfcowell, @manusis, @mgummelt -- do you have any thoughts on this issue? Thanks!

With best regards,

Elmer

**Comment by kyle221b** [comment]
Anybody have thoughts on this?

**Review by JenniferMah — APPROVED** [review]
